### PR TITLE
8303431: [JVMCI] libgraal annotation API

### PR DESCRIPTION
**Pull request by dougxc**
This PR extends JVMCI with new API (`jdk.vm.ci.meta.Annotated`) for accessing annotations. The main differences from `java.lang.reflect.AnnotatedElement` are:
* Each `Annotated` method explicitly specifies the annotation type(s) for which it wants annotation data. That is, there is no direct equivalent of `AnnotatedElement.getAnnotations()`.
* Annotation data is returned in a map-like object (of type `jdk.vm.ci.meta.AnnotationData`) instead of in an `Annotation` object. This works better for libgraal as it avoids the need for annotation types to be loaded and included in libgraal.
 
To demonstrate the new API, here's an example in terms `java.lang.reflect.AnnotatedElement` (which `ResolvedJavaType` implements):
```
    ResolvedJavaMethod method = ...;
    ExplodeLoop a = method.getAnnotation(ExplodeLoop.class);
    return switch (a.kind()) {
        case FULL_UNROLL -> LoopExplosionKind.FULL_UNROLL;
        case FULL_UNROLL_UNTIL_RETURN -> LoopExplosionKind.FULL_UNROLL_UNTIL_RETURN;
        ...
    }
```

The same code using the new API:

```
    ResolvedJavaMethod method = ...;
    ResolvedJavaType explodeLoopType = ...;
    AnnotationData a = method.getAnnotationDataFor(explodeLoopType);
    return switch (a.getEnum("kind").getName()) {
        case "FULL_UNROLL" -> LoopExplosionKind.FULL_UNROLL;
        case "FULL_UNROLL_UNTIL_RETURN" -> LoopExplosionKind.FULL_UNROLL_UNTIL_RETURN;
        ...
    }
```

The implementation relies on new methods in `jdk.internal.vm.VMSupport` for parsing annotations and serializing/deserializing to/from a byte array. This allows the annotation data to be passed from the HotSpot heap to the libgraal heap.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303431](https://bugs.openjdk.org/browse/JDK-8303431): [JVMCI] libgraal annotation API

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12810/head:pull/12810` \
`$ git checkout pull/12810`

Update a local copy of the PR: \
`$ git checkout pull/12810` \
`$ git pull https://git.openjdk.org/jdk pull/12810/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12810`

View PR using the GUI difftool: \
`$ git pr show -t 12810`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12810.diff">https://git.openjdk.org/jdk/pull/12810.diff</a>

</details>

### CODE REVIEW

File: test/jdk/jdk/internal/vm/TestTranslatedException.java [hunk 1]
```diff
@@ -58,15 +58,6 @@ public void encodeDecodeTest() throws Exception {
...
     }
-    @SuppressWarnings("unchecked")
-    @Test
-    public void encodeDecodeTest2() throws Exception {
-        Throwable throwable = new ExceptionInInitializerError(new InvocationTargetException(new Untranslatable("test exception", new NullPointerException()), "invoke"));
-        for (int i = 0; i < 10; i++) {
-            throwable = new ExceptionInInitializerError(new InvocationTargetException(new RuntimeException(String.valueOf(i), throwable), "invoke"));
-        }
-        encodeDecode(throwable);
-    }
```
Comment (vnkozlov) [inline test/jdk/jdk/internal/vm/TestTranslatedException.java:61]: Why this was removed?

**Review by vnkozlov — APPROVED** [review]
I looked on Hotspot, JVMCI and VMSupport.java changes. But you need  to ask Tom to look on JVMCI changes in details. And someone from core-libs who familiar with Annotations have to comment on your implementation in general because I am not expert.

File: test/jdk/jdk/internal/vm/TestTranslatedException.java (hunk 1 above)
Comment (dougxc) [reply test/jdk/jdk/internal/vm/TestTranslatedException.java:61]: Because it does exactly the same thing as `encodeDecodeTest`. It should have been cleaned up in the original PR that introduced this test.

**Review by tkrodriguez — APPROVED** [review]
The JVMCI changes look ok to me.

File: src/hotspot/share/jvmci/jvmciCompilerToVM.cpp [hunk 2]
```diff
@@ -2622,24 +2622,120 @@ C2V_VMENTRY_NULL(jobject, asReflectionExecutable, (JNIEnv* env, jobject, ARGUMEN
...
+                         CHECK_NULL);
+
+  oop res = result.get_oop();
+  if (JVMCIENV->is_hotspot()) {
+    return (jbyteArray) JNIHandles::make_local(THREAD, res);
+  }
+
+  typeArrayOop ba = typeArrayOop(res);
+  int ba_len = ba->length();
+  if (ba_len <= 256) {
```
[the remark this answers is not in the document]
Comment (dougxc) [reply src/hotspot/share/jvmci/jvmciCompilerToVM.cpp:2699]: Ok, good point. I'll remove the optimization.

**Comment by dougxc** [comment]
@jddarcy would you be able to help review this PR? Based on git log, you are knowledgeable in `sun.reflect.annotation`. If not, please suggest who else I can reach out to for a review.

**Comment by jddarcy** [comment]
> @jddarcy would you be able to help review this PR? Based on git log, you are knowledgeable in `sun.reflect.annotation`. If not, please suggest who else I can reach out to for a review.

I can take a look at this, but it will be at least a few days before I can swap it in. I've also worked on the annotation-reading API in javax.lang.model.

I assume https://bugs.openjdk.org/browse/JDK-8303431 recounts the motivation behind this change?

**Comment by dougxc** [comment]
> I assume https://bugs.openjdk.org/browse/JDK-8303431 recounts the motivation behind this change?

Yes, it does. Thanks in advance.

File: src/java.base/share/classes/jdk/internal/vm/VMSupport.java [hunk 3]
```diff
@@ -167,4 +179,404 @@ public static int encodeThrowable(Throwable throwable, long buffer, int bufferSi
...
+        return encodeAnnotations(annotations.values());
+    }
+
+    /**
+     * Encodes a list of annotations to a byte array. The byte array can be decoded with {@link #decodeAnnotations(byte[], AnnotationDecoder)}.
+     */
+    public static byte[] encodeAnnotations(Collection<Annotation> annotations) {
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream(128);
+            try(DataOutputStream dos = new DataOutputStream(baos)) {
```
Comment (turbanoff) [inline src/java.base/share/classes/jdk/internal/vm/VMSupport.java:237]: nit
```suggestion
            try (DataOutputStream dos = new DataOutputStream(baos)) {
```

File: src/java.base/share/classes/jdk/internal/vm/VMSupport.java [hunk 4]
```diff
@@ -167,4 +179,404 @@ public static int encodeThrowable(Throwable throwable, long buffer, int bufferSi
...
+                superClass = superClass.getSuperclass();
+            }
+        }
+        return encodeAnnotations(annotations.values());
+    }
+
+    /**
+     * Encodes a list of annotations to a byte array. The byte array can be decoded with {@link #decodeAnnotations(byte[], AnnotationDecoder)}.
+     */
+    public static byte[] encodeAnnotations(Collection<Annotation> annotations) {
```
Comment (jddarcy) [inline src/java.base/share/classes/jdk/internal/vm/VMSupport.java:234]: I don't think it matters much in this use case, but it looks like  encodeAnnotations could be changed to take a List rather than a Collection, as the comment implies.

File: src/java.base/share/classes/jdk/internal/vm/VMSupport.java [hunk 5]
```diff
@@ -167,4 +179,404 @@ public static int encodeThrowable(Throwable throwable, long buffer, int bufferSi
...
+    /**
+     * Helper for {@link #decodeAnnotations(byte[], AnnotationDecoder)} to convert a byte
+     * array (ostensibly produced by {@link VMSupport#encodeAnnotations}) into objects.
+     *
+     * @param <T> type to which a type name is {@linkplain #resolveType(String) resolved}
+     * @param <A> type of the object representing a decoded annotation
+     * @param <E> type of the object representing a decoded enum constant
+     * @param <X> type of the object representing a decoded error
+     */
+    public interface AnnotationDecoder<T, A, E, X> {
```
Comment (jddarcy) [inline src/java.base/share/classes/jdk/internal/vm/VMSupport.java:419]: I think it would be better to include some bound on the type parameters to better capture their intention
A extends java.lang.Annotatoin
E extends java.lang.Enum
etc.

**Comment by jddarcy** [comment]
A few higher-level comments:

From the long-term perspective, it is likely that the set of kinds of elements that can occur in an annotation will be expanded, for example, method references are a repeated request. Easing future maintenance to gives more inter-source linkage in this situation and error handling for this case in the libgraal code would be prudent IMO.

The java.lang.reflect.AnnotatedElement API (https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/lang/reflect/AnnotatedElement.html) defines different ways an annotation can be affiliated with an element:

"The terms directly present, indirectly present, present, and associated are used throughout this interface to describe precisely which annotations are returned by methods: "

tl;dr these terms relate to which of inheriting annotations and looking through repeated annotations the methods do on behalf of the caller. I think the methods should phrase their operations in terms of these concepts, such as "Construct the annotations present..." since inheritance is taken into account.

File: src/java.base/share/classes/jdk/internal/vm/VMSupport.java (hunk 5 above)
Comment (dougxc) [reply src/java.base/share/classes/jdk/internal/vm/VMSupport.java:419]: These types are *alternatives* to `java.lang.Annotation`, `java.lang.Enum` etc. That's the primary motivation for this PR, i.e. to be able to represent annotations without having to reify them as `java.lang.Annotation` objects.

File: src/java.base/share/classes/jdk/internal/vm/VMSupport.java (hunk 4 above)
Comment (dougxc) [reply src/java.base/share/classes/jdk/internal/vm/VMSupport.java:234]: Just above (line 228) you can see a call to this method where the argument comes from `Map.values()` whose type is `Collection` so I'd prefer to leave it as is rather than have to convert the argument to a `List`.

Comment (dougxc) [reply src/java.base/share/classes/jdk/internal/vm/VMSupport.java:234]: Just above (line 228) you can see a call to this method where the argument comes from `Map.values()` whose type is `Collection` so I'd prefer to leave it as is rather than have to convert the argument to a `List`.

**Comment by dougxc** [comment]
> From the long-term perspective, it is likely that the set of kinds of elements that can occur in an annotation will be expanded, for example, method references are a repeated request. Easing future maintenance to gives more inter-source linkage in this situation and error handling for this case in the libgraal code would be prudent IMO.

I'm not sure what you're suggesting in terms of how I should update this PR. Do you mean `AnnotationData.get` needs to somehow be more flexible? If so, could you please give a concrete example of what you're after.

File: src/java.base/share/classes/jdk/internal/vm/VMSupport.java (hunk 4 above)
Comment (jddarcy) [reply src/java.base/share/classes/jdk/internal/vm/VMSupport.java:234]: In that case, I think the comment on line 232 should be updated to not say "list".

**Comment by jddarcy** [comment]
> > From the long-term perspective, it is likely that the set of kinds of elements that can occur in an annotation will be expanded, for example, method references are a repeated request. Easing future maintenance to gives more inter-source linkage in this situation and error handling for this case in the libgraal code would be prudent IMO.
> 
> I'm not sure what you're suggesting in terms of how I should update this PR. Do you mean `AnnotationData.get` needs to somehow be more flexible? If so, could you please give a concrete example of what you're after.

Let me explain my concerns in more detail.

We have (at least) two separate annotations implementations in the JDK, one in javac for compile-time and other in core reflection for runtime. Due to various technical constraints, the implementations are necessarily separate (although the annotation objects constructed by javac do also implement the java.lang.annotation.Annotation interface also used by core reflection).

This work is proposing to add another partial implementation to satisfy other technical goals, reusing portions of the existing core reflection machinery.

If at some point the universe of objects that can be encoded as annotation is expanded, e..g method literals as mentioned previously, it is well-understood that core reflection and javac will need to be updated. I think it would be relatively easy to overlook the need to make corresponding updates to libgraal API and all regression tests might still pass.

As a concrete suggestion, if such an unknown annotation was handed over to libgraal, I think the code should reject it (AssertionError("unexpected annotation component") etc.) in hope of the omission getting corrected sooner. Also, leaving some bread crumb comments to future maintainers of core reflection annotations in that implementation package would be helpful too, e.g. "// used by libgraal".

HTH

**Review by jddarcy — APPROVED** [review]

**Comment by vnkozlov** [comment]
@dougxc I see next 2 JVMCI tests failed when run `-XX:TypeProfileLevel=222` on all platforms (our stress testing):
```
compiler/jvmci/jdk.vm.ci.runtime.test/src/jdk/vm/ci/runtime/test/TestResolvedJavaMethod.java
compiler/jvmci/jdk.vm.ci.runtime.test/src/jdk/vm/ci/runtime/test/TestResolvedJavaType.java

#  Internal Error (/workspace/open/src/hotspot/cpu/x86/macroAssembler_x86.cpp:829), pid=2430194, tid=2430218
#  fatal error: DEBUG MESSAGE: exact klass and actual klass differ

Current thread (0x00007f9ff8460480):  JavaThread "MainThread" [_thread_in_Java, id=2430218, stack(0x00007f9fe06b7000,0x00007f9fe07b8000)]

Stack: [0x00007f9fe06b7000,0x00007f9fe07b8000],  sp=0x00007f9fe07b44e0,  free space=1013k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x12b8d35]  MacroAssembler::debug64(char*, long, long*)+0x45  (macroAssembler_x86.cpp:829)
J 1935 c1 sun.reflect.annotation.AnnotationParser.parseAnnotation2(Ljava/nio/ByteBuffer;Ljdk/internal/reflect/ConstantPool;Ljava/lang/Class;Z[Ljava/lang/Class;)Ljava/lang/annotation/Annotation; java.base@21-internal (275 bytes) @ 0x00007f9fe10d4cc6 [0x00007f9fe10d4740+0x0000000000000586]

[error occurred during error reporting (printing native stack (with source info))
```

**Comment by dougxc** [comment]
I can reproduce this locally as well but really don't know where to start in terms of debugging this. Can you please provide hints as to what may cause this failure C1 compiled code?
